### PR TITLE
Adds Character Loadout Tab to preferences (with just a small handful of items to start)

### PR DESCRIPTION
**Pull request by MrMelbert**
## About The Pull Request

Adds a Character Loadout Tab to the preference menu

This tab lets you pick items to start the round with

image

This also has some additional mechanics, such as being able to recolor colorable items, rename certain items (such as plushies), set item skins (such as the pride pin)

image

## Why It's Good For The Game

This has been headcoder sanctioned for some time, just no one did it. So here we are.

Allows players to add some additional customization to their characters. Keeps us from cluttering the quirks list with quirks that do nothing but grants items. 

## Changelog

:cl: Melbert
add: Character Loadouts
del: Pride Pin quirk (it's in the Loadout menu now)
/:cl:

### CODE REVIEW

**Comment by SomeoneYouProbablyKnow** [comment]
Yippe

**Comment by MrMelbert** [comment]
In this pr
```
Current items
- Glasses
(Glasses available to nearsighted quirk, glasses available in costume vendor, blindfold, eyepatch)

- Hats
(Beanie, cap of all colors, berets, fedoreas, cat ears, rabbit ears, some costume hats (bandana), top hat, bowler hat, ushanka, bear pelt, rose, wig)

- Inhands
(Cane, white cane, briefcase, locked briefcase)

- Neck
(Scarves, neckties)

- Pocket items
(Plushies (all kinds), lipstick, card binder, kotahi deck, wizoff deck, dice)
```

**Comment by MrMelbert** [comment]
I need to make the pride pin quirk migrate

**Comment by sylvia-from-fulp-station** [comment]
how much do i have to paypal MSO for roundstart tacticool skirt

**Comment by SomeoneYouProbablyKnow** [comment]
> how much do i have to paypal MSO for roundstart tacticool skirt

15 points worth of negative quirks should unlock the tacticool skirt

**Comment by Gaxeer** [comment]
What do you think about changing button bg color from blue to something more neutral?

**Comment by Jacquerel** [comment]
Not totally your call but before we merge this we probably want a headcoder approved "what does and doesn't go in loadouts" thing to point at for future PRs 
I know Oranges has a red line for changing uniform for instance, good to have it written down somewhere

**Comment by DrAmazing343** [comment]
HERO

**Comment by Jezzared** [comment]
Colorist can also be moved to this system, can't it?

**Comment by Dawnseer** [comment]
Is there a way to limit how much you can pick? Or can I get a vest, tie, briefcase and glasses at the same time for that business ready look?

In any case, glad to see it finally happen

**Comment by Sadboysuss** [comment]
Another Melbert W

**Comment by SimplySmiley123** [comment]
Finally, a way to start with a scarf.

**Comment by Sadboysuss** [comment]
if you can select accessories, make sure it doesn't conflict with the station scarves trait

**Comment by Sadboysuss** [comment]
also potentially - the cost of these items should be the cost of them in the vendors and use your round start money, but that may be annoying for different jobs and is entirely up to you.

**Comment by Gaxeer** [comment]
Do you plan to add loadout restrictions by job and `loadout point` to have limit of loadout taken and make some items more expensive etc. ?

**Comment by MatrixOne500** [comment]
Yes, this is amazing. I really hope we can start with a scarf of the preferred color or to change the color of shoes! great change

File: tgui/packages/tgui/interfaces/PreferencesMenu/LoadoutManager.tsx
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/PreferencesMenu/LoadoutManager.tsx:69]: No need to cast these, this is inferred

File: tgui/packages/tgui/interfaces/PreferencesMenu/LoadoutManager.tsx
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/PreferencesMenu/LoadoutManager.tsx:159]:
```suggestion
  const { act } = useBackend();
```
Does nothing here

File: tgui/packages/tgui/interfaces/PreferencesMenu/LoadoutManager.tsx
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/PreferencesMenu/LoadoutManager.tsx:189]:
```suggestion
  const { buttons } = item;
```
Prefer destructuring pls

File: tgui/packages/tgui/interfaces/PreferencesMenu/LoadoutManager.tsx
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/PreferencesMenu/LoadoutManager.tsx:172]:
```suggestion
      tooltipPosition="bottom-start"
```
Minor

File: tgui/packages/tgui/interfaces/PreferencesMenu/LoadoutManager.tsx
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/PreferencesMenu/LoadoutManager.tsx:276]: @stylemistake had some negative opinions on this, consider an a>b sort rather than localecompare

**Review by jlsnow301 — COMMENTED** [review]
Not too shabby on the ui side

**Comment by Loafin34** [comment]
my god.... add coats please. if you could do coupons that'd be awesome, i loved the station trait where assistants were able to spec into beekeeping or skateboarding.

**Comment by DaCoolBoss** [comment]
Moth and lizard plushies will need to be removed as progtot objectives if they're available here. People are lame enough to take them for a potential in-game advantage.

**Comment by SomeoneYouProbablyKnow** [comment]
Can also just blacklist loudout plushies from those objectives

**Comment by Mickyan** [comment]
I do have a bit of an issue with wigs being available as I think it invalidates the smooth headed quirk without all the fun interactions it comes with

**Comment by Jezzared** [comment]
The loadout is meant to remove the tedium of buying your outfit every round, with the added bonus of small personalization like named plushies.
If someone wears a wig every round without using the bald quirk, then removing it from loadouts won't stop them, it just makes the process more annoying for them.

**Comment by Livrah** [comment]
I like this. especially with the greyscale options.
But one of the reasons I chose this particular server is the lack of 100 things I need/can customise. idk why it wasn't allowed for some time but it was a positive in some way.

**Comment by Admiral-Gigaton** [comment]
the loadouttening

**Comment by SomeoneYouProbablyKnow** [comment]
> But one of the reasons I chose this particular server is the lack of 100 things I need/can customise. idk why it wasn't allowed for some time but it was a positive in some way.

There is something to say about people actually looking uniformed and uniform but for that you'd need to crack down on clothing vendors and clothing options. Not much good is coming from these shiftstart chores except wasted time.

**Comment by Admiral-Gigaton** [comment]
if you can, make sure to add a job lock, so john assistant cant get cappy tan's sweet alternate uniform (if ever added to this system, that is)

**Comment by Jacquerel** [comment]
> if you can, make sure to add a job lock, so john assistant cant get cappy tan's sweet alternate uniform (if ever added to this system, that is)

There is a headcoder block on uniforms being part of this system, you will not be able to not spawn in your job outfit.

**Comment by SomeoneYouProbablyKnow** [comment]
Can you add the wheelchair to this? I want to be able be get a roundstart wheelchair from time to time and then midrounf just start walking normally to mess with people

**Comment by MrMelbert** [comment]
Update 
image

File: tgui/packages/tgui/interfaces/PreferencesMenu/loadout/index.tsx
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/PreferencesMenu/loadout/index.tsx:182]:
```suggestion
const typepathToLoadoutItem = (
```
only components that return jsx get pascal case

as for your comment below it's probably faster with object but not enough of an issue to be necessary. the bundler will also mangle this btw. usememo is another route too if you were worried about it

https://react.dev/reference/react/useMemo

**Review by jlsnow301 — APPROVED** [review]
looks good

**Comment by Mothblocks** [comment]
Going to read through this sometime this week

File: tgui/packages/tgui/interfaces/PreferencesMenu/loadout/index.tsx
Comment (Mothblocks) [reply tgui/packages/tgui/interfaces/PreferencesMenu/loadout/index.tsx:182]: yep

**Comment by Mothblocks** [comment]
@MrMelbert can you update op to the new figma design?

File: code/modules/admin/verbs/mapping.dm
Comment (Mothblocks) [inline code/modules/admin/verbs/mapping.dm:212]: named params? and ditch null

File: code/modules/client/preferences_savefile.dm
Comment (Mothblocks) [inline code/modules/client/preferences_savefile.dm:108]: Not the right solution to this, too easy to just update migrate_quirks_to_loadout and nothing else.

I think instead this should be something like `migrate_quirk_to_loadout("Pride Pin", whatever)`, and then we write them out manually here since I can't imagine we ever do more than a handful at a time

File: code/modules/client/preferences/migrations/quirk_loadout_migration.dm
Comment (Mothblocks) [inline code/modules/client/preferences/migrations/quirk_loadout_migration.dm:20]: Not a useful comment

File: code/modules/loadout/categories/accessories.dm
Comment (Mothblocks) [inline code/modules/loadout/categories/accessories.dm:65]: Default values having actual behavior is spooky and I remember it causing problems from time to time, let's avoid this

File: code/modules/loadout/categories/accessories.dm
Comment (Mothblocks) [inline code/modules/loadout/categories/accessories.dm:72]: Are these ever not already bools?

File: code/modules/loadout/categories/glasses.dm
Comment (Mothblocks) [inline code/modules/loadout/categories/glasses.dm:1]: The file path itself makes this obvious IMO

File: code/modules/loadout/categories/glasses.dm
Comment (Mothblocks) [inline code/modules/loadout/categories/glasses.dm:8]: These should either be defines in one central place or even better, `/datum/loadout_category/head::tab_order + 1`

File: code/modules/loadout/loadout_categories.dm
Comment (Mothblocks) [inline code/modules/loadout/loadout_categories.dm:44]: Feels like temporary doesn't really help anything here

File: code/modules/loadout/loadout_categories.dm
Comment (Mothblocks) [inline code/modules/loadout/loadout_categories.dm:57]:
```suggestion
	sortTim(formatted_list, /proc/cmp_assoc_list_name) // Alphabetizing
```

File: code/modules/loadout/loadout_items.dm
Comment (Mothblocks) [inline code/modules/loadout/loadout_items.dm:34]: Nit: Feels like this should be at the end so the user provided stuff comes first

File: code/modules/loadout/loadout_items.dm
Comment (Mothblocks) [inline code/modules/loadout/loadout_items.dm:65]:
```suggestion
```

File: code/modules/loadout/loadout_items.dm
Comment (Mothblocks) [inline code/modules/loadout/loadout_items.dm:72]: These two comments are self evident from the code

File: code/modules/loadout/loadout_items.dm
Comment (Mothblocks) [inline code/modules/loadout/loadout_items.dm:87]: Can you check what the total cost of /datum/loadout_item/New in the init_profiler is? I'm suspicious of the performance of this especially as we start to add more items

File: code/modules/loadout/loadout_items.dm
Comment (Mothblocks) [inline code/modules/loadout/loadout_items.dm:125]: I don't know if anyone remembers this, but you should basically always return TRUE (here and in the other early returns) in the event of an unexpected error. It doesn't mean success--it means update the UI, and it might fix whatever misprediction forced it to get into the state where it sends an invalid action.

File: code/modules/loadout/loadout_items.dm
Comment (Mothblocks) [inline code/modules/loadout/loadout_items.dm:130]: I think it should either silently fail or close the last menu because otherwise if you're clicking the color prompt a couple times because it's the server initializing and it's laggy, you're just gonna get more annoyed with all the popups.

File: code/modules/loadout/loadout_items.dm
Comment (Mothblocks) [inline code/modules/loadout/loadout_items.dm:180]: Should this be "the"?

File: code/modules/loadout/loadout_items.dm
Comment (Mothblocks) [inline code/modules/loadout/loadout_items.dm:335]:
```suggestion
```

File: code/modules/loadout/loadout_items.dm
Comment (Mothblocks) [inline code/modules/loadout/loadout_items.dm:293]: Should probably doc all these

File: tgui/packages/tgui/interfaces/PreferencesMenu/loadout/item_display.tsx
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/PreferencesMenu/loadout/item_display.tsx:54]: assert?

File: tgui/packages/tgui/interfaces/PreferencesMenu/loadout/item_display.tsx
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/PreferencesMenu/loadout/item_display.tsx:56]: camelCase in js

File: tgui/packages/tgui/interfaces/PreferencesMenu/loadout/item_display.tsx
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/PreferencesMenu/loadout/item_display.tsx:1]: Name: PascalCase for component files generally--the only reason some are odd in prefs menu is because they correlate to a dynamic DM string

File: tgui/packages/tgui/interfaces/PreferencesMenu/loadout/item_display.tsx
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/PreferencesMenu/loadout/item_display.tsx:135]: Why is this a function?

File: tgui/packages/tgui/interfaces/PreferencesMenu/loadout/item_display.tsx
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/PreferencesMenu/loadout/item_display.tsx:144]: Don't we have a `createSearch` helper for this exact thing? We use it in orbit menu if you want to look

**Comment by CopperCrow9** [comment]
quick question: Would it be a good idea to include certain racial cosmetics in this menu too? I'm talking like the different lizard tails and the felinid ears.

File: code/modules/loadout/categories/accessories.dm
Comment (MrMelbert) [reply code/modules/loadout/categories/accessories.dm:72]: I put it there in the event it's `null` from being unset, I don't think it matters ultimately but I didn't want to mess with it

File: code/modules/loadout/loadout_items.dm
Comment (MrMelbert) [reply code/modules/loadout/loadout_items.dm:34]: From past experience people will continue to tack user provided stuff after wherever you put it without a care

File: code/modules/loadout/loadout_items.dm
Comment (MrMelbert) [reply code/modules/loadout/loadout_items.dm:87]:
```
/datum/loadout_item/New          0.002        0.002        0.002        0.002           91
/datum/loadout_category/New      0.000        0.002        0.002        0.000            6
```

File: tgui/packages/tgui/interfaces/PreferencesMenu/loadout/index.tsx
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/PreferencesMenu/loadout/index.tsx:63]: clever

File: tgui/packages/tgui/interfaces/PreferencesMenu/loadout/index.tsx
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/PreferencesMenu/loadout/index.tsx:277]: keys are required to be on the topmost child elements within map, which in this case is the empty fragment, not loadoutselecteditem. this could actually cause issues, so you'd need to move the `key` prop to the empty <>. import and use `Fragment` to do this.

File: tgui/packages/tgui/interfaces/PreferencesMenu/loadout/index.tsx
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/PreferencesMenu/loadout/index.tsx:307]:
```suggestion
      <Stack vertical fill>
```

File: tgui/packages/tgui/interfaces/PreferencesMenu/loadout/ItemDisplay.tsx
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/PreferencesMenu/loadout/ItemDisplay.tsx:133]: this is given, createSearch does this by default
```suggestion
    currentSearch,
```

File: tgui/packages/tgui/interfaces/PreferencesMenu/loadout/index.tsx
Comment (MrMelbert) [reply tgui/packages/tgui/interfaces/PreferencesMenu/loadout/index.tsx:63]: Credit to MB for it

**Pull request merged**
